draft-ietf-httpbis-compression-dictionary.md

```diff
@@ -81,39 +81,39 @@ be used as a dictionary for future requests for URLs that match the rules
 specified in the Use-As-Dictionary response header.
 
 The Use-As-Dictionary response header is a Structured Field
-{{STRUCTURED-FIELDS}} sf-dictionary with values for "match", "match-search",
-"match-dest", "ttl", "id", and "type".
+{{STRUCTURED-FIELDS}} sf-dictionary with values for "match", "match-dest",
+"ttl", "id", and "type".
 
 ### match
 
 The "match" value of the Use-As-Dictionary header is a sf-string value
-that provides the "pathname" of a URLPattern
-(https://urlpattern.spec.whatwg.org/#dom-urlpatterninit-pathname).
-
-The "match" is a {{URL}} path relative to the full request URL of the
-dictionary.  The request URL for the dictionary itself is used as the "baseURL"
-for constructing the URLPattern (https://urlpattern.spec.whatwg.org/) that
-is used for matching the dictionary to relevant requests when running
-{{dictionary-url-matching}}.
-
-The URLPattern used for request matching does not support regular expressions
-(https://urlpattern.spec.whatwg.org/#token-type-regexp) in the "match".
+that provides the URLPattern to use for request matching
+(https://urlpattern.spec.whatwg.org/).
+
+The URLPattern used for matching has some restrictions:
+* Regular expressions are not supported
+(https://urlpattern.spec.whatwg.org/#token-type-regexp).
+* The match pattern MUST be for the same {{Origin}} as the dictionary.
+
+The following algorithm will return TRUE for a valid match pattern and FALSE
+for an invalid pattern that MUST NOT be used:
+
+1. Let MATCH be the value of "match" for the given dictionary.
+1. Let URL be the URL of the dictionary request.
+1. Let PATTERN be a URLPattern constructed by setting input=MATCH,
+and baseURL=BASEURL (https://urlpattern.spec.whatwg.org/).
+1. If PATTERN parsed a regexp token then return FALSE
+(https://urlpattern.spec.whatwg.org/#token-type-regexp).
+1. If the "hash" attribute of PATTERN is a non-empty string then return FALSE.
+1. Let SRC be a URLPattern constructed by setting input=BASEURL.
+1. If any of the "protocol", "username", "password", "hostname", or "port"
+attributes of PATTERN do not match the value of the same attribute in SRC then
+return FALSE.
+1. Return True.
```

Comment:
Allowing for invalid patterns to be stored can be pretty expensive and it would be better if we can fail any cross-origin patterns at the time of storage rather than at the time of matching. The cost comes from having to store the dictionary resource (can be multiple megabytes) as well as evaluating the pattern on each request.

The origin checking matters here at the HTTP level because of cross-request compression-based encryption attacks.

Reply:
Is this difference between storing vs. not observable, such that it needs to be specified?

Reply:
Good point, thanks.  I cleaned it up to do the same-origin check at match time instead of dictionary storage time and use the URLs themselves for the check instead of relying on the dictionary.